netaddr.go

```diff
@@ -846,3 +871,422 @@ func (p *IPPrefix) UnmarshalText(text []byte) error {
+	return !r.From.IsZero() && !r.To.IsZero() &&
+		r.From.Is4() == r.To.Is4() &&
+		!r.To.Less(r.From)
+}
+
+// ip16 represents a mutable IP address, either IPv4 (in IPv6-mapped
+// form) or IPv6.
+type ip16 [16]byte
+
+// bitSet reports whether the given bit in the address is set.
+// (bit 0 is the most significant bit in ip[0]; bit 127 is last)
+func (ip ip16) bitSet(bit uint8) bool {
+	i, s := bit/8, 7-(bit%8)
+	return ip[i]&(1<<s) != 0
+}
+
+func (ip *ip16) set(bit uint8) {
+	i, s := bit/8, 7-(bit%8)
+	ip[i] |= 1 << s
+}
+
+func (ip *ip16) clear(bit uint8) {
+	i, s := bit/8, 7-(bit%8)
+	ip[i] &^= 1 << s
+}
+
+// lastWithBitZero returns a copy of ip with the given bit
+// cleared and the following all set.
+func (ip ip16) lastWithBitZero(bit uint8) ip16 {
+	ip.clear(bit)
+	for ; bit < 128; bit++ {
+		ip.set(bit)
+	}
+	return ip
+}
+
+// firstWithBitOne returns a copy of ip with the given bit
+// set and the following all cleared.
+func (ip ip16) firstWithBitOne(bit uint8) ip16 {
+	ip.set(bit)
+	for ; bit < 128; bit++ {
+		ip.clear(bit)
+	}
+	return ip
+}
+
+// prefixMaker returns a address-family-corrected IPPrefix from ip16 and bits,
+// where the input bits is always in the IPv6-mapped form for IPv4 addresses.
+type prefixMaker func(ip16 ip16, bits uint8) IPPrefix
+
+// Prefixes returns the set of IPPrefix entries that covers r.
+//
+// If either of r's bounds are invalid, in the wrong order, or if
+// they're of different address families, then Prefixes returns nil.
+func (r IPRange) Prefixes() []IPPrefix {
+	if !r.Valid() {
+		return nil
+	}
+	var makePrefix prefixMaker
+	if r.From.Is4() {
+		makePrefix = func(ip16 ip16, bits uint8) IPPrefix {
+			return IPPrefix{IPFrom16([16]byte(ip16)), bits - 12*8}
+		}
+	} else {
+		makePrefix = func(ip16 ip16, bits uint8) IPPrefix {
+			return IPPrefix{IPv6Raw([16]byte(ip16)), bits}
+		}
+	}
+	a16, b16 := ip16(r.From.As16()), ip16(r.To.As16())
+	return appendRangePrefixes(nil, makePrefix, a16, b16)
+}
+
+func appendRangePrefixes(dst []IPPrefix, makePrefix prefixMaker, a16, b16 ip16) []IPPrefix {
+	common := uint8(0)
+	for common < 128 && a16.bitSet(common) == b16.bitSet(common) {
+		common++
+	}
+	// See whether a16 and b16, after their common shared bits, end
+	// in all zero bits or all one bits, respectively.
+	aAllZero, bAllSet := true, true
+	for i := common; i < 128; i++ {
+		if a16.bitSet(i) {
+			aAllZero = false
+			break
+		}
+	}
+	for i := common; i < 128; i++ {
+		if !b16.bitSet(i) {
+			bAllSet = false
+			break
+		}
+	}
+	if aAllZero && bAllSet {
+		// a16 to b16 represents a whole range, like 10.50.0.0/16.
+		// (a16 being 10.50.0.0 and b16 being 10.50.255.255)
+		return append(dst, makePrefix(a16, common))
+	}
+	// Otherwise recursively do both halves.
+	dst = appendRangePrefixes(dst, makePrefix, a16, a16.lastWithBitZero(common+1))
+	dst = appendRangePrefixes(dst, makePrefix, b16.firstWithBitOne(common+1), b16)
+	return dst
+}
+
+func addOne(a []byte, i int) bool {
```

Comment:
This is fine as-is, but an alternative is to accept just a byte slice and use the end of the slice to locate i. Then reslice to recurse (or better make iterative). Then call sites look like `addOne(a[12:15])` instead of `addOne(a[12:], 3)`, which strikes me as a bit clearer.

Reply:
It's a pity math/big.nat isn't exported, 'cause it'd be perfect for a lot of these operations. Inc, AndNot, etc.